File: LaboratoryWork/Semester_2/Laboratory_7/Laboratory_2/PHP/task2.php
<?php
// Включим отображение ошибок
error_reporting(E_ALL);
ini_set("display_errors", 1);


function print_bottles(string $bottle, int $count) {
    /* Выводит бутылки на экран. */
    for ($i = 0; $i < $count; $i++) {
        if ($i % 3 == 0) {
            print("\n");
        }
        print($bottle . " ");
    }
}


function drink_simulation(int $n, int $m) {
    /* Показывает все шаги питья. */
    $full = $n;
    $empty = 0;
    $total_drunk = 0;
    $iterations = 0;

    $full_bottle = "●";
    $empty_bottle = "○";

    // Выведем дано
    print("Дано");
    print_bottles($full_bottle, $full);
    print("\n\n");

    while (true) {
        // Шаг 1: выпить все полные бутылки
        if ($full > 0) {
            $total_drunk += $full;
            $empty += $full;
            $full = 0;
            $iterations++;

            // Выведем все выпитые бутылки
            print($iterations . " шаг");
            print_bottles($empty_bottle, $empty);
            print("\n\n");

            if ($empty < $m) {  // Условие завершения
                break;
            }

            // Шаг 2: обмен пустых бутылок на полные
            $full = (int)($empty / $m);
            $empty = $empty % $m;
            $iterations++;

            // Выведем бутылки после обмена
            print($iterations . " шаг");
            print_bottles($full_bottle, $full);
            print_bottles($empty_bottle, $empty);
            print("\n\n");
        }
    }

    return [$total_drunk, $iterations];
}


function main() {
    // Используем sscanf для извлечения двух чисел
    sscanf(readline("Введите два числа через пробел: "), "%d %d", $n, $m);

    print("Input: " . $n . " " . $m);

    $result = drink_simulation($n, $m);

    $total = $result[0];
    $iters = $result[1];

    print("Output (total, iters): " . $total . " " . $iters);
}

main();
?>

File: LaboratoryWork/Semester_2/Laboratory_7/Laboratory_2/C++/task2.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;


void print_bottles(string bottle, int count) {
    /* Выводит бутылки на экран. */
    for (int i = 0; i < count; i++) {
        if (i % 3 == 0) cout << endl;
        cout << bottle << " ";
    }
}


pair<int, int> drink_simulation(int n, int m) {
    /* Показывает все шаги питья. */
    int full = n;
    int empty = 0;
    int total_drunk = 0;
    int iterations = 0;

    string full_bottle  = "●";
    string empty_bottle = "○";

    // Выведем дано
    cout << "Дано";
    print_bottles(full_bottle, full);
    cout << endl << endl;
    
    while (true) {
        // Шаг 1: выпить все полные бутылки
        if (full > 0) {
            total_drunk += full;
            empty += full;
            full = 0;
            iterations++;

            // Выведем все выпитые бутылки
            cout << iterations << " шаг";
            print_bottles(empty_bottle, empty);
            cout << endl << endl;

            if (empty < m) break;  // условие завершения
            
            // Шаг 2: обмен пустых бутылок на полные
            full = empty / m;
            empty = empty % m;
            iterations++;

            // Выведем бутылки после обмена 
            cout << iterations << " шаг";
            print_bottles(full_bottle, full);
            print_bottles(empty_bottle, empty);
            cout << endl << endl;
        }
    }

    return {total_drunk, iterations};
}


int main(void) {
    int n, m;
    cout << "Введите количество полных бутылок и количество бутылок, которых достаточно для обмена: ";
    cin >> n >> m;

    cout << "Input: " << n << " " << m << endl;

    pair<int, int> result = drink_simulation(n, m);

    int total = result.first;
    int iters = result.second;

    cout << "Output (total, iters): " << total << " " << iters << endl;
}